File: iot/amqp/header/impl/AMQPTransfer.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.HeaderCode as HeaderCode
import iot.amqp.avps.ReceiveCode as ReceiveCode
import iot.amqp.avps.SectionCode as SectionCode
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.header.api.HeaderFactoryOutcome as HeaderFactoryOutcome
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVList as TLVList
import iot.amqp.wrappers.AMQPMessageFormat as AMQPMessageFormat

class amqpTransfer():
    def __init__(self,code,doff,type,channel,handle,deliveryId,deliveryTag,messageFormat,settled,more,rcvSettleMode,state,resume,aborted,batchable,sections):
        if code is not None:
            self.code = code
        else:
            self.code = HeaderCode.headerCode.getValueByKey('TRANSFER')
        if doff is not None:
            self.doff = doff
        else:
            self.doff = 2
        if type is not None:
            self.type = type
        else:
            self.type = 0
        if channel is not None:
            self.channel = channel
        else:
            self.channel = 0

        self.handle = handle
        self.deliveryId = deliveryId
        self.deliveryTag = deliveryTag
        self.messageFormat = messageFormat
        self.settled = settled
        self.more = more
        self.rcvSettleMode = rcvSettleMode
        self.state = state
        self.resume = resume
        self.aborted = aborted
        self.batchable = batchable
        self.sections = sections

    def setDoff(self,doff):
        self.doff = doff

    def getDoff(self):
        return self.doff

    def getType(self):
        return self.type

    def setType(self, type):
        self.type = type

    def getChannel(self):
        return self.channel

    def setChannel(self, channel):
        self.channel = channel

    def getCode(self):
        return self.code

    def toArgumentsList(self):
        list = TLVList.tlvList(None,None)
        wrapper = AMQPWrapper.amqpWrapper()
        if self.handle == None:
            print("Transfer header's handle can't be null")
        list.addElement(0,wrapper.wrap(self.handle))
        if self.deliveryId is not None:
            list.addElement(1, wrapper.wrap(self.deliveryId))
        if self.deliveryTag is not None:
            list.addElement(2, wrapper.wrap(self.deliveryTag))
        if self.messageFormat is not None:
            list.addElement(3, wrapper.wrap(self.messageFormat.encode()))
        if self.settled is not None:
            list.addElement(4, wrapper.wrap(self.settled))
        if self.more is not None:
            list.addElement(5, wrapper.wrap(self.more))
        if self.rcvSettleMode is not None:
            list.addElement(6, wrapper.wrap(self.rcvSettleMode.value))
        if self.state is not None:
            list.addElement(7, wrapper.wrap(self.state.toArgumentsList()))
        if self.resume is not None:
            list.addElement(8, wrapper.wrap(self.resume))
        if self.aborted is not None:
            list.addElement(9, wrapper.wrap(self.aborted))
        if self.batchable is not None:
            list.addElement(10, wrapper.wrap(self.batchable))

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(AMQPType.amqpType.getValueByKey('SMALL_ULONG'), self.code.value))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        size = len(list.getList())
        if size == 0:
            print("Received malformed Transfer header: handle can't be null")
        if size > 11:
            print('Received malformed Transfer header. Invalid number of arguments: ' + str(size))
        if size > 0:
            element = list.getList()[0]
            if element is None and not element.isNull():
                print("Received malformed Transfer header: handle can't be null")
            self.handle = unwrapper.unwrapUInt(element)
        if size > 1:
            element = list.getList()[1]
            if element is not None and not element.isNull():
                self.deliveryId = unwrapper.unwrapUInt(element)
        if size > 2:
            element = list.getList()[2]
            if element is not None and not element.isNull():
                self.deliveryTag = unwrapper.unwrapBinary(element)
        if size > 3:
            element = list.getList()[3]
            if element is not None and not element.isNull():
                self.messageFormat = AMQPMessageFormat.amqpMessageFormat(unwrapper.unwrapUInt(element))
        if size > 4:
            element = list.getList()[4]
            if element is not None and not element.isNull():
                self.settled = unwrapper.unwrapBool(element)
        if size > 5:
            element = list.getList()[5]
            if element is not None and not element.isNull():
                self.more = unwrapper.unwrapBool(element)
        if size > 6:
            element = list.getList()[6]
            if element is not None and not element.isNull():
                self.rcvSettleMode = ReceiveCode.receiveCode(unwrapper.unwrapUByte(element))
        if size > 7:
            element = list.getList()[7]
            if element is not None and not element.isNull():
                code = element.getCode()
                if code not in (AMQPType.amqpType.getValueByKey('LIST_0'),AMQPType.amqpType.getValueByKey('LIST_8'),AMQPType.amqpType.getValueByKey('LIST_32')):
                    print("Expected type 'STATE' - received: " + str(element.getCode()))
                self.state = HeaderFactoryOutcome.headerFactoryOutcome.getState(element)
                self.state.fromArgumentsList(element)
        if size > 8:
            element = list.getList()[8]
            if element is not None and not element.isNull():
                self.resume = unwrapper.unwrapBool(element)
        if size > 9:
            element = list.getList()[9]
            if element is not None and not element.isNull():
                self.aborted = unwrapper.unwrapBool(element)
        if size > 10:
            element = list.getList()[10]
            if element is not None and not element.isNull():
                self.batchable = unwrapper.unwrapBool(element)

    def toString(self):
        return "AMQPTransfer [handle=" + str(self.handle) + ", deliveryId=" + str(self.deliveryId) + ", deliveryTag=" + str(self.deliveryTag) + ", messageFormat=" + str(self.messageFormat) + ", settled=" + str(self.settled) + ", more=" + str(self.more) + ", rcvSettleMode=" + str(self.rcvSettleMode) + ", state=" + str(self.state) + ", resume=" + str(self.resume) + ", aborted=" + str(self.aborted) + ", batchable=" + str(self.batchable) + ", sections=" + str(self.sections) + ", code=" + str(self.code) + ", doff=" + str(self.doff) + ", type=" + str(self.type) + ", channel=" + str(self.channel) + "]"

    def setHandle(self, handle):
        self.handle = handle

    def getHandle(self):
        return self.handle

    def setDeliveryId(self, deliveryId):
        self.deliveryId = deliveryId

    def getDeliveryId(self):
        return self.deliveryId

    def setDeliveryTag(self, deliveryTag):
        self.deliveryTag = deliveryTag

    def getDeliveryTag(self):
        return self.deliveryTag

    def setMessageFormat(self, messageFormat):
        self.messageFormat = messageFormat

    def getMessageFormat(self):
        return self.messageFormat

    def setSettled(self, settled):
        self.settled = settled

    def getSettled(self):
        return self.settled

    def setMore(self, more):
        self.more = more

    def getMore(self):
        return self.more

    def setRcvSettleMode(self, rcvSettleMode):
        self.rcvSettleMode = rcvSettleMode

    def getRcvSettleMode(self):
        return self.rcvSettleMode

    def setState(self, state):
        self.state = state

    def getState(self):
        return self.state

    def setResume(self, resume):
        self.resume = resume

    def getResume(self):
        return self.resume

    def setAborted(self, aborted):
        self.aborted = aborted

    def getAborted(self):
        return self.aborted

    def setBatchable(self, batchable):
        self.batchable = batchable

    def getBatchable(self):
        return self.batchable

    def setSections(self, sections):
        self.sections = sections

    def getSections(self):
        return self.sections

    def getHeader(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('HEADER')]
        else:
            return None

    def getDeliveryAnnotations(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('DELIVERY_ANNOTATIONS')]
        else:
            return None

    def getMessageAnnotations(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('MESSAGE_ANNOTATIONS')]
        else:
            return None

    def getProperties(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('PROPERTIES')]
        else:
            return None

    def getApplicationProperties(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('APPLICATION_PROPERTIES')]
        else:
            return None

    def getData(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('DATA')]
        else:
            return None

    def getSequence(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('SEQUENCE')]
        else:
            return None

    def getValue(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('VALUE')]
        else:
            return None

    def getFooter(self):
        if self.sections is not None:
            return self.sections[SectionCode.sectionCode.getValueByKey('FOOTER')]
        else:
            return None

File: IoTbrokerClient/src/iot-protocols/amqp/classes/headeramqp/amqptransfer.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqptransfer.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvfixed.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvnull.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpfactory.h"
#include "iot-protocols/amqp/classes/section/amqpdata.h"

AMQPTransfer::AMQPTransfer() : AMQPHeader(new AMQPHeaderCode(AMQP_TRANSFER_HEADER_CODE))
{
    this->sections = NULL;
    this->handle = NULL;
    this->deliveryId = NULL;
    this->deliveryTag = malloc(sizeof(char));
    this->messageFormat = NULL;
    this->settled = NULL;
    this->more = NULL;
    this->rcvSettleMode = NULL;
    this->state = NULL;
    this->resume = NULL;
    this->aborted = NULL;
    this->batchable = NULL;
    this->sections = NULL;
}

int AMQPTransfer::getLength()
{
    int length = 8;
    AMQPTLVList *arguments = this->arguments();
    length += arguments->getLength();

    std::list<AMQPSection *> array = this->sections.values();
    foreach (AMQPSection *item, array) {
        length += item->getValue()->getLength();
    }
    return length;
}

int AMQPTransfer::getType()
{
    return AMQP_TRANSFER_HEADER_CODE;
}

AMQPTLVList *AMQPTransfer::arguments()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (this->handle == NULL) {
        printf("AMQPTransfer::arguments::handle");
        return new AMQPTLVList();
    }
    list->addElementWithIndex(0, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->handle)));

    if (this->deliveryId != NULL) {
        list->addElementWithIndex(1, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->deliveryId)));
    }
    if (this->deliveryTag[0] != 0) {
        list->addElementWithIndex(2, AMQPWrapper::wrapBinary(this->deliveryTag));
    }
    if (this->messageFormat != NULL) {
        list->addElementWithIndex(3, AMQPWrapper::wrapUInt(this->messageFormat->encode()));
    }
    if (this->settled != NULL) {
        list->addElementWithIndex(4, AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(this->settled)));
    }
    if (this->more != NULL) {
        list->addElementWithIndex(5, AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(this->more)));
    }
    if (this->rcvSettleMode != NULL) {
        list->addElementWithIndex(6, AMQPWrapper::wrapUByte(this->rcvSettleMode->getValue()));
    }
    if (this->state != NULL) {
        list->addElementWithIndex(7, this->state->getList());
    }
    if (this->resume != NULL) {
        list->addElementWithIndex(8, AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(this->resume)));
    }
    if (this->aborted != NULL) {
        list->addElementWithIndex(9, AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(this->aborted)));
    }
    if (this->batchable != NULL) {
        list->addElementWithIndex(10, AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(this->batchable)));
    }
    unsigned char * data = malloc(sizeof(char));
    data[0] = this->code->getValue();
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);
    return list;
}

void AMQPTransfer::fillArguments(AMQPTLVList *list)
{
    int size = list->getList().count();

    if (size == 0) {
        printf("AMQPTransfer::fillArguments::size == 0");
        return;
    }
    if (size > 11) {
        printf("AMQPTransfer::fillArguments::wrong size");
        return;
    }

    if (size > 0) {
        TLVAMQP *element = list->getList()[0];
        if (element->isNull()) {
            printf("AMQPTransfer::fillArguments::handle");
            return;
        }
        this->handle = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
    }
    if (size > 1) {
        TLVAMQP *element = list->getList()[1];
        if (!element->isNull()) {
            this->deliveryId = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
        }
    }
    if (size > 2) {
        TLVAMQP *element = list->getList()[2];
        if (!element->isNull()) {
            this->deliveryTag = AMQPUnwrapper::unwrapBinary(element);
        }
    }
    if (size > 3) {
        TLVAMQP *element = list->getList()[3];
        if (!element->isNull()) {
            this->messageFormat = new AMQPMessageFormat(AMQPUnwrapper::unwrapUInt(element));
        }
    }
    if (size > 4) {
        TLVAMQP *element = list->getList()[4];
        if (!element->isNull()) {
            this->settled = AMQPSimpleType::boolToVariant(AMQPUnwrapper::unwrapBool(element));
        }
    }
    if (size > 5) {
        TLVAMQP *element = list->getList()[5];
        if (!element->isNull()) {
            this->more = AMQPSimpleType::boolToVariant(AMQPUnwrapper::unwrapBool(element));
        }
    }
    if (size > 6) {
        TLVAMQP *element = list->getList()[6];
        if (!element->isNull()) {
            this->rcvSettleMode = new AMQPReceiverSettleMode(AMQPUnwrapper::unwrapUByte(element));
        }
    }
    if (size > 7) {
        TLVAMQP *element = list->getList()[7];
        if (!element->isNull()) {
            if (element->getTypeValue() != AMQP_LIST0_TYPE && element->getTypeValue() != AMQP_LIST8_TYPE && element->getTypeValue() != AMQP_LIST32_TYPE) {
                printf("AMQPTransfer::fillArguments");
                return;
            }
            this->state = AMQPFactory::state((AMQPTLVList *)element);
            this->state->fill((AMQPTLVList *)element);
        }
    }
    if (size > 8) {
        TLVAMQP *element = list->getList()[8];
        if (!element->isNull()) {
            this->resume = AMQPSimpleType::boolToVariant(AMQPUnwrapper::unwrapBool(element));
        }
    }
    if (size > 9) {
        TLVAMQP *element = list->getList()[9];
        if (!element->isNull()) {
            this->aborted = AMQPSimpleType::boolToVariant(AMQPUnwrapper::unwrapBool(element));
        }
    }
    if (size > 10) {
        TLVAMQP *element = list->getList()[10];
        if (!element->isNull()) {
            this->batchable = AMQPSimpleType::boolToVariant(AMQPUnwrapper::unwrapBool(element));
        }
    }
}

AMQPSection *AMQPTransfer::getHeader()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_HEADER_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getDeliveryAnnotations()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_DELIVERY_ANNOTATIONS_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getMessageAnnotations()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_MESSAGE_ANNOTATIONS_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getProperties()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_PROPERTIES_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getApplicationProperties()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_APPLICATION_PROPERTIES_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getData()
{
    AMQPData *data = NULL;

    foreach (AMQPSection *item, this->sections.values()) {
        if (qobject_cast<AMQPData *>(item) != NULL) {
            data = (AMQPData *)item;
        }
    }
    return data;
}

AMQPSection *AMQPTransfer::getSequence()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_SEQUENCE_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getValue()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_VALUE_SECTION_CODE);
    return this->sections.value(sectionCode);
}

AMQPSection *AMQPTransfer::getFooter()
{
    AMQPSectionCode *sectionCode = new AMQPSectionCode(AMQP_FOOTER_SECTION_CODE);
    return this->sections.value(sectionCode);
}

void AMQPTransfer::addSections(std::list<AMQPSection *> array)
{
    foreach (AMQPSection *item, array) {
        this->sections.insert(item->getCode(), item);
    }
}

JsonVariant *AMQPTransfer::getHandle() const
{
    return handle;
}

void AMQPTransfer::setHandle(JsonVariant *value)
{
    handle = value;
}

JsonVariant *AMQPTransfer::getDeliveryId() const
{
    return deliveryId;
}

void AMQPTransfer::setDeliveryId(JsonVariant *value)
{
    deliveryId = value;
}

unsigned char * AMQPTransfer::getDeliveryTag() const
{
    return deliveryTag;
}

void AMQPTransfer::setDeliveryTag(const unsigned char &value)
{
    deliveryTag = value;
}

AMQPMessageFormat *AMQPTransfer::getMessageFormat() const
{
    return messageFormat;
}

void AMQPTransfer::setMessageFormat(AMQPMessageFormat *value)
{
    messageFormat = value;
}

JsonVariant *AMQPTransfer::getSettled() const
{
    return settled;
}

void AMQPTransfer::setSettled(JsonVariant *value)
{
    settled = value;
}

JsonVariant *AMQPTransfer::getMore() const
{
    return more;
}

void AMQPTransfer::setMore(JsonVariant *value)
{
    more = value;
}

AMQPReceiverSettleMode *AMQPTransfer::getRcvSettleMode() const
{
    return rcvSettleMode;
}

void AMQPTransfer::setRcvSettleMode(AMQPReceiverSettleMode *value)
{
    rcvSettleMode = value;
}

AMQPState *AMQPTransfer::getState() const
{
    return state;
}

void AMQPTransfer::setState(AMQPState *value)
{
    state = value;
}

JsonVariant *AMQPTransfer::getResume() const
{
    return resume;
}

void AMQPTransfer::setResume(JsonVariant *value)
{
    resume = value;
}

JsonVariant *AMQPTransfer::getAborted() const
{
    return aborted;
}

void AMQPTransfer::setAborted(JsonVariant *value)
{
    aborted = value;
}

JsonVariant *AMQPTransfer::getBatchable() const
{
    return batchable;
}

void AMQPTransfer::setBatchable(JsonVariant *value)
{
    batchable = value;
}

std::map<AMQPSectionCode *, AMQPSection *> AMQPTransfer::getSections() const
{
    return sections;
}

void AMQPTransfer::setSections(const std::map<AMQPSectionCode *, AMQPSection *> &value)
{
    sections = value;
}
